File: androidApp/src/main/java/me/vishnu/tracktracker/android/screens/carScreen.kt
package me.vishnu.tracktracker.android.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import me.vishnu.tracktracker.shared.models.titleDisplay
import me.vishnu.tracktracker.shared.stores.Loop
import me.vishnu.tracktracker.shared.stores.welcome.*

@Composable
fun CarScreen(carScreenStore: CarScreenStore, carScreenEffectHandler: CarScreenEffectHandler) {
  val (loopState, dispatch) = remember {
    Loop(
      store = carScreenStore,
      effectHandler = carScreenEffectHandler,
      startEffects = { setOf(CarScreenEffects.LoadInitialData) },
    )
  }

  CarScreen(
    state = loopState.collectAsState(initial = CarScreenModel()).value,
    dispatch = dispatch
  )
}

@Composable
private fun CarScreen(state: CarScreenModel, dispatch: (CarScreenEvents) -> Unit) {
  Scaffold(
    topBar = { TopAppBar(title = { Text("Track Tracker") }) },
    floatingActionButton = {
      if (!state.addCarMode) {
        FloatingActionButton(onClick = {
          dispatch(CarScreenEvents.AddCar)
        }) { Icon(Icons.Default.Add, contentDescription = "Add Car Button") }
      }
    }
  ) {
    if (state.addCarMode) {
      CarScreenAddCar(state.selectableYears, dispatch)
    } else {
      CarScreenDisplay(state)
    }
  }
}

@Composable
private fun CarScreenAddCar(selectableYears: List<Int>, eventCallback: (CarScreenEvents) -> Unit) {
  val year = remember { mutableStateOf(selectableYears.first()) }
  val yearDisplay = if (year.value == 0) "" else year.value.toString()
  val make = remember { mutableStateOf("") }
  val model = remember { mutableStateOf("") }
  val trim = remember { mutableStateOf("") }
  val nickname = remember { mutableStateOf("") }

  val yearIsValid = selectableYears.contains(year.value)

  val yearDropdownExpanded = remember { mutableStateOf(false) }
  Column(
    modifier = Modifier
      .fillMaxWidth()
      .padding(16.dp)
  ) {

    OutlinedTextField(
      modifier = Modifier.fillMaxWidth(),
      keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
      value = yearDisplay,
      isError = !yearIsValid,
      onValueChange = { newValue: String ->
        if (newValue.isNotEmpty()) {
          year.value = newValue.toInt()
        } else {
          year.value = 0
        }
      },
      label = { Text("Year") },
      trailingIcon = {
        Box {
          Icon(
            modifier = Modifier.clickable { yearDropdownExpanded.value = true },
            imageVector = Icons.Default.ArrowDropDown,
            contentDescription = "Dropdown for year selection"
          )

          DropdownMenu(
            expanded = yearDropdownExpanded.value,
            onDismissRequest = { yearDropdownExpanded.value = false }) {
            selectableYears.forEach { yearToDisplay ->
              DropdownMenuItem(onClick = {
                year.value = yearToDisplay
                yearDropdownExpanded.value = false
              }) {
                Text("$yearToDisplay")
              }
            }
          }

        }
      }
    )
    CarInputField(label = "Make", value = make)
    CarInputField(label = "Model", value = model)
    CarInputField(label = "Trim", value = trim)
    CarInputField(label = "Nickname", value = nickname)

    Row {
      Button(onClick = { eventCallback(CarScreenEvents.DismissAddCarDialog) }) {
        Text("Cancel")
      }

      Button(onClick = {
        eventCallback(
          CarScreenEvents.CreateCar(
            year = year.value,
            make = make.value,
            model = model.value,
            trim = trim.value,
            nickname = nickname.value
          )
        )
      }) {
        Text("Submit")
      }
    }
  }
}

@Composable
private fun CarInputField(label: String, value: MutableState<String>) {
  TextField(
    modifier = Modifier.fillMaxWidth(),
    value = value.value,
    onValueChange = { value.value = it },
    label = {
      Text(label)
    }
  )
}

@Composable
private fun CarScreenDisplay(state: CarScreenModel) {
  if (state.cars.isEmpty()) {
    Text("add some cars!")
  } else {
    LazyColumn {
      items(state.cars) { car ->
        Column(modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)) {
          Text(car.titleDisplay(), style = MaterialTheme.typography.h6)
          if (car.nickname != null) {
            Text(car.nickname!!, style = MaterialTheme.typography.subtitle1)
          }
        }
      }
    }
  }
}

File: iosApp/iosApp/CarScreen.swift
import SwiftUI
import shared


func carScreen(component: AppComponent) -> some View{
  typealias ObservableCarScreen = ObservableStore<CarScreenStore, CarScreenModel, CarScreenEvents, CarScreenEffects>
  let obsStore: ObservableCarScreen
  let carScreenComponent = InjectCarScreenComponent(parent: component)
  
  let store = carScreenComponent.carScreenStore
  let effectHandler = carScreenComponent.carScreenEffectHandler
  obsStore = ObservableCarScreen(
    store: store,
    state: CarScreenModel.Companion.shared.defaultModel(),
    stateWatcher: store.watchState(),
    sideEffectWatcher: store.watchSideEffect()
  )
  
  _ = Loop<CarScreenModel, CarScreenEvents, CarScreenEffects, CarScreenEffectHandler>(
    store: store,
    effectHandler: effectHandler,
    startEffects: {
      let initEffects : Set = [CarScreenEffects.LoadInitialData.shared]
      return initEffects
    },
    eventSources: [Flow]())
  
  return CarScreen().environmentObject(obsStore)
}

struct CarScreen: ConnectedView {
  typealias S = CarScreenStore
  typealias M = CarScreenModel
  typealias E = CarScreenEvents
  typealias F = CarScreenEffects
  
  struct Props {
    let addCarMode: Bool
    let cars: [UiCar]
    let selectableYears: [Int]
    let onAddCar: () -> Void
    let onDiscardCreateCar: () -> Void
    let onCreateCar: (Int, String, String, String?, String?) -> Void
  }
  
  func map(state: CarScreenModel, dispatch: @escaping DispatchFunction<CarScreenEvents>) -> Props {
    return Props(
      addCarMode: state.addCarMode,
      cars: state.cars,
      selectableYears: state.selectableYears.map({ kInt in
        kInt.intValue
      }),
      onAddCar: { dispatch(CarScreenEvents.AddCar.shared)},
      onDiscardCreateCar: { dispatch(CarScreenEvents.DismissAddCarDialog.shared)},
      onCreateCar: { year, make, model, trim, nickname in
        dispatch(CarScreenEvents.CreateCar(
          year: Int32(year), make: make, model: model, trim: trim, nickname: nickname
        ))
      }
    )
  }
  
  func body(props: Props) -> some View {
    NavigationView {
      VStack {
        if (props.addCarMode) {
          AddCarView(selectableYears: props.selectableYears, onDiscard: props.onDiscardCreateCar, onSubmit: props.onCreateCar)
        }
        else {
          if (props.cars.isEmpty) {
            addCarButton(onAddCar: props.onAddCar)
          }
          else {
            ScrollView {
              LazyVStack(alignment: .leading) {
                ForEach(props.cars, id: \.self) { car in
                  Text("\(car.titleDisplay())")
                    .font(.title3)
                  if let nickname = car.nickname {
                    Text(nickname)
                      .font(.subheadline)
                  }
                }
              }
            }
            addCarButton(onAddCar: props.onAddCar)
            
          }
        }
      }
      .navigationTitle("Track Tracker")
    }
  }
  
  func addCarButton(onAddCar: @escaping () -> Void) -> some View {
    Button("Add a car") {
      onAddCar()
    }
  }
}

struct AddCarView: View {
  @State private var year: Int
  @State private var make: String = ""
  @State private var model: String = ""
  @State private var trim: String = ""
  @State private var nickname: String = ""
  
  let onDiscard: () -> Void
  let onSubmit: (Int, String, String, String?, String?) -> Void
  let selectableYears: [Int]
  
  init(selectableYears: [Int], onDiscard: @escaping () -> Void, onSubmit: @escaping (Int, String, String, String?, String?) -> Void ) {
    self.selectableYears = selectableYears
    self.year = self.selectableYears.first ?? 2021
    self.onDiscard = onDiscard
    self.onSubmit = onSubmit
  }
  
  var body: some View {
    VStack {
      Form {
        Picker("Year", selection: $year) {
          ForEach(selectableYears, id: \.self) {
            Text(String($0))
          }
        }
        .pickerStyle(DefaultPickerStyle())
        TextField("Make", text: $make)
        TextField("Model", text: $model)
        TextField("trim", text: $trim)
        TextField("nickname", text: $nickname)
      }
      HStack(spacing: 24) {
        Button("Discard") {
          onDiscard()
        }
        Button("Submit") {
          onSubmit(year, make, model, trim, nickname)
        }
      }
    }
  }
}
